jwql/website/apps/jwql/monitor_pages/monitor_readnoise_bokeh.py

```diff
@@ -59,102 +72,128 @@ def load_data(self):
+
+            amp_plot.xaxis.axis_label = 'Date'
+            amp_plot.yaxis.axis_label = 'Mean Readnoise [DN]'
+
+            self.amp_plots.append(amp_plot)
+
+    def plot_readnoise_difference_image(self):
+        """Updates the readnoise difference image"""
 
         # Update the readnoise difference image and histogram, if data exists
-        if len(self.query_results) != 0:
-            # Get the most recent data; the entries were sorted by time when
-            # loading the database, so the last entry will always be the most recent.
-            diff_image_png = self.query_results[-1].readnoise_diff_image
+
+        self.diff_image_plot = figure(title='Readnoise Difference (most recent dark - pipeline reffile)',
+                                      height=500, width=500, sizing_mode='scale_width')
+
+        if len(self.db.query_results) != 0:
+            diff_image_png = self.db.query_results[-1].readnoise_diff_image
             diff_image_png = os.path.join('/static', '/'.join(diff_image_png.split('/')[-6:]))
-            diff_image_n = np.array(self.query_results[-1].diff_image_n)
-            diff_image_bin_centers = np.array(self.query_results[-1].diff_image_bin_centers)
-
-            # Update the readnoise difference image and histogram
-            self.refs['readnoise_diff_image'].image_url(url=[diff_image_png], x=0, y=0, w=2048, h=2048, anchor="bottom_left")
-            self.refs['diff_hist_source'].data = {'n': diff_image_n,
-                                                  'bin_centers': diff_image_bin_centers}
-            self.refs['diff_hist_xr'].start = diff_image_bin_centers.min()
-            self.refs['diff_hist_xr'].end = diff_image_bin_centers.max()
-            self.refs['diff_hist_yr'].start = diff_image_n.min()
-            self.refs['diff_hist_yr'].end = diff_image_n.max() + diff_image_n.max() * 0.05
-
-        # Update the readnoise difference image style
-        self.refs['readnoise_diff_image'].xaxis.visible = False
-        self.refs['readnoise_diff_image'].yaxis.visible = False
-        self.refs['readnoise_diff_image'].xgrid.grid_line_color = None
-        self.refs['readnoise_diff_image'].ygrid.grid_line_color = None
-        self.refs['readnoise_diff_image'].title.text_font_size = '22px'
-        self.refs['readnoise_diff_image'].title.align = 'center'
+            self.diff_image_plot.image_url(url=[diff_image_png], x=0, y=0, w=2048, h=2048, anchor="bottom_left")
+            self.diff_image_plot.xaxis.visible = False
+            self.diff_image_plot.yaxis.visible = False
+            self.diff_image_plot.xgrid.grid_line_color = None
+            self.diff_image_plot.ygrid.grid_line_color = None
+            self.diff_image_plot.title.text_font_size = '22px'
+            self.diff_image_plot.title.align = 'center'
+
+    def plot_readnoise_histogram(self):
+        """Updates the readnoise histogram"""
+
+        diff_image_n = np.array(self.db.query_results[-1].diff_image_n)
```

Comment:
This looks like it would crash if self.db.query_results is empty

Reply:
I went ahead and added some boolean logic here to hopefully avoid this problem

```
        if len(self.db.query_results) != 0:
            diff_image_n = np.array(self.db.query_results[-1].diff_image_n)
            diff_image_bin_centers = np.array(self.db.query_results[-1].diff_image_bin_centers)
        else:
            diff_image_n = np.array(list())
            diff_image_bin_centers = np.array(list())
```

Reply:
Nice!